### ISSUE TITLE
Decide on a recursive attribute set tree flattening function

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#237776]
**Project description**

All of the PRs below try to add a function that recurses into a tree of attrsets.
Clearly there's a need. What needs to be done is decide on the name and behavior of the function.
The name should fit the established naming conventions. The behavior should probably be biased towards keeping it simple and "composable".
Some discussion seems necessary, hence a new thread in this issue to discuss the new function's details.

 - https://github.com/NixOS/nixpkgs/pull/206965
 - https://github.com/NixOS/nixpkgs/pull/202517
 - https://github.com/NixOS/nixpkgs/pull/221608
 - https://github.com/NixOS/nixpkgs/pull/234615

**Metadata**

* homepage URL:
* source URL: `./lib`
* license: mit
* platforms: nix

**Comment by figsoda** [comment]
maybe we can start with `concatMapAttrsWith`, something like [this](https://github.com/nix-community/haumea/blob/7b93825345b989d70374f9ad6b24d1ed75dfba9c/src/transformers/__utils/concatMapAttrsWith.nix)

**Comment by roberth** [comment]
> maybe we can start with `concatMapAttrsWith`, something like [this](https://github.com/nix-community/haumea/blob/7b93825345b989d70374f9ad6b24d1ed75dfba9c/src/transformers/__utils/concatMapAttrsWith.nix)

Is that `merge` function (used in the linked definition) a recursive merge? Recursive merging without a conditional for identifying leaves will construct the weirdest Frankenstein-esque values if used incorrectly by accident, or used "correctly" on user inputs that can't be validated, which in my opinion is sufficient reason to discourage its indirect use through functions such as the one proposed in this issue.

**Comment by figsoda** [comment]
> Is that `merge` function (used in the linked definition) a recursive merge?

It can be if you pass in a recursive merge function, is just a more generalized version of the `concatMapAttrs` in nixpkgs

`concatMapAttrsWith (x: y: x // y)` would be equivalent to `concatMapAttrs`, and `concatMapAttrsWith recursiveUpdate` would be recursive merge

**Comment by roberth** [comment]
We have rather many possible combinations of recursion, filtering, mapping, and ways to recombine information into a single attribute name.

Picking specific combinations of these strategies and giving a name to them isn't particularly helpful, especially if done in an ad hoc way (without much consideration for naming conventions that would effectively only serve to refer to those multiple strategies in ways that won't be as good as an actual programmatic decomposition of the problem). The [Fairbairn threshold] suggests that we wouldn't be making any progress towards making the library and programs more understandable that way (at best).

I think the problem _can_ be decomposed at the call site, taking the following shape:

```nix
someTraversal
  maybeSomeParameter
  (callbackParams:
    {
      ${attributeName} = callbackParams.value;
    })
```

Now this pattern doesn't look great, but it does convey all the information in a systematic manner, and unlike an semi-incoherent set of names, this is more like a useful skill that I believe is easy to learn, and can be taught (e.g. in the docs) with a few examples.

What are the choices for each strategy?

- Choice of the traversal
  - non-recursive: `concatMapAttrs`, with callback parameters `name: value:`
  - recursive: `concatMapAttrsRecursive`, with callback parameters `attrPath: value`

Note: we may want to change `concat` to `merge`; see "alternatives" in https://github.com/NixOS/nix/issues/11887

- Choice of filtering
  - non-recursive: `concatMapAttrs` with or without `optionalAttrs` in the callback
  - recursive: `concatMapAttrsRecursive` or `concatMapAttrsRecursiveWith` (where `maybeSomeParameter` exists and determines which nodes qualify as `value` for the callback).
  
- Choice of naming (`attributeName` in the pattern)
  - `last attrPath`
  - `concatStringsSep "." attrPath`

- Bonus: you can return 0 to many attributes

Example: in the context of flakes, convert `legacyPackages` to a `packages` compatible flat attrset.

```nix
concatMapAttrsRecursiveWith (_path: isDerivation) (attrPath: value: {
  ${concatStringsSep "." attrPath} = value;
})
```

Would `flattenAttrsRecursiveWithSep "." isDerivation` be better? Maybe, because I'm familiar with some of the other function names, but this creates a sprawl of functions that each have to be learned or looked up, whereas the example can be understood by knowing the basic functions. What's more, it will be easier to make most modifications, because you can edit right there how the attribute names are constructed, or how the values are filtered.

So I think a function like `flattenAttrsRecursiveWithSep` does not contribute to making the library easier to use.
Or maybe I'm wrong, but I think it would be prudent to start with a better primitive like `concatMapAttrsRecursiveWith`, and only develop conveniences after that.

I'd happily accept a `flattenLegacyPackages {}` instead of a `flattenAttrsRecursiveWithSep`, because it is a proper abstraction that is named in its own terminology, and not a function that only seems to exists for the purpose of typing 10-20% fewer characters while obfuscating what really happens, despite all the good intentions.

tl;dr: start with `concatMapAttrsRecursiveWith` and `concatMapAttrsRecursive`?

(btw the `x` + `xWith` pattern is well established in this family of functions, so that's cheap; not all names should weigh equally when judging their cost/benefit. Having a good system reduces the cost a bit.)

Related:
- [concatMapAttrs](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.attrsets.concatMapAttrs)
- [foldAttrs](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.attrsets.foldAttrs)
- [zipAttrsWith](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.attrsets.zipAttrsWith)

[Fairbairn threshold]: https://wiki.haskell.org/Fairbairn_threshold

**Comment by hsjobeki** [comment]
My personal liking is towards "collect" 
Especially because it is short and memorizable.
(Say hello: fairbairn threshold) 

I think the Initial implementation of collect wasnt generic enough. 

All of the linked PRs can be acheived using collect' from one of the PRs and sometimes wrapping it with listToAttrs. 

i think the Interface is good and Kind of Matches your Idea. I dislike that overly precise name (concatMapAttrsRecursiveUntil) because its to long and hard to type.


maybe we can also find a way that avoids creating an intermediate list for performance reasons.

**Comment by roberth** [comment]
> overly precise name (concatMapAttrsRecursiveUntil)

I agree. Was trying to stick with the naming system, but that may not be necessary.


> find a way that avoids creating an intermediate list

Achievable with `concatMapAttrs`, although the implementation isn't currently free of lists, but I propose to fix that:
- https://github.com/NixOS/nix/issues/11887
- perhaps no surprise considering even https://github.com/NixOS/nix/issues/11888 is yet to be done

**Comment by roberth** [comment]
> > `concatMapAttrsRecursive`

Transforming the leaves of a tree is not a particularly "recursive" computation from a user perspective.

**Comment by removed-user** [comment]
Partly related?

Using this to map legacyPackages to flat packages?; certainly think it could be cleaner, but taking this as example gives me some idea's/suggestions for the final function here.

```nix
# flatten.nix
{ lib }:
let
  # Helper to find how many 'prev' links exist deep inside an attrset
  findMaxPrevDepth = attrs:
    if builtins.hasAttr "prev" attrs && builtins.isAttrs attrs.prev then
      1 + findMaxPrevDepth attrs.prev
    else
      0;

  # baseName: The original package root (e.g., "tinycc")
  # currentDepth: How many 'prev' layers deep we currently are
  # maxDepth: The total number of 'prev' layers found for this package
  flattenAttrs = baseName: currentDepth: maxDepth: attrs:
    lib.concatMapAttrs (name: value:
      let
        isPrev = name == "prev";
        
        # Calculate parameters for the next iteration
        nextDepth = if isPrev then currentDepth + 1 else currentDepth;
        
        # If we are at the top level, discover the max depth for this package chain
        nextMaxDepth = if baseName == "" && builtins.isAttrs value then 
          findMaxPrevDepth value 
        else 
          maxDepth;

        # Track the name of the tool/file inside the attribute set
        nextBaseName = 
          if baseName == "" then name
          else if isPrev then baseName  # Keep the root name, don't append "-stageX" yet
          else "${baseName}/${name}";    # For other nested attributes like sub-tools

        # Calculate the chronological bootstrap stage
        stageNum = nextMaxDepth - nextDepth;
        
        # Format the final name cleanly only when we hit a derivation 
# (if derivation is a target/match condition)
        makeFinalKey = if nextDepth == 0 then nextBaseName else "${nextBaseName}-stage${toString stageNum}";
      in
      if lib.isDerivation value then
        { "${makeFinalKey}" = value; }
      else if builtins.isAttrs value then
        flattenAttrs nextBaseName nextDepth nextMaxDepth value
      else
        {}
    ) attrs;
in
flattenAttrs "" 0 0
```

By "could be cleaner", I'm thinking about what would make this more generally useful, and what "component args", and return values the optimally reusable generic recursion function would take and return.

**Args given**:

Starting point: the attrset to start recursing from.

Recurse-Condition: A function determining whether we choose to continue traversing the current subtree. Which is checked - after checking if we even can continue recursing.

Matcher(s)/EffectCondition: A function, or attrset of functions, used to determine whether we apply our effect(s) to the given/current node.

Thinking for best reuse, this could basically accept all checks and actions as "plugin" functions. 

Take paired attrsets of named functions, and effects, so that any time a condition is matched we can trigger the paired effect on the node that triggered it.

Effects/Actions/Transform: The effect or action to take on chosen attrs/those that return true for a given condition.

Each of these defaults to null/no-op, so irrelevant logic is never evaluated.

That way, it could practically behave as a "lightweight" but flexible function, which is practically adaptable to ALL attribute recursion tasks. 

You give a plugin function/adapter for each of the various tasks that recursion is used for.

At each node along the way, we evaluate checks, determine whether we can recurse any further, whether we should, what effect conditions have matched, apply any matching effects/transformations, and then continue.
